File: electrobun_backend/src/bun/lib/parse-env.ts
export type ParsedEnvEntry = { key: string; value: string }

export class EnvParseError extends Error {
  line: number

  constructor(message: string, line: number) {
    super(`${message} (line ${line})`)
    this.line = line
  }
}

/**
 * Parse a `.env`-style file into key/value pairs.
 *
 * Supported syntax:
 * - `KEY=VALUE` and `export KEY=VALUE`
 * - Full-line comments (`# ...`)
 * - Trailing comments after an unquoted value (`KEY=value # note`)
 * - Single- or double-quoted values (`KEY="value"`)
 * - Blank lines (ignored)
 *
 * Throws `EnvParseError` when a content line is not a valid assignment.
 */
export function parseEnvFile(content: string): ParsedEnvEntry[] {
  const entries: ParsedEnvEntry[] = []
  const lines = content.split(/\r?\n/)

  for (let i = 0; i < lines.length; i++) {
    let line = lines[i]!
    let trimmed = line.trim()

    if (!trimmed || trimmed.startsWith("#")) continue
    if (trimmed.startsWith("export ")) {
      line = trimmed.slice("export ".length)
      trimmed = line.trim()
    }

    const eq = trimmed.indexOf("=")
    if (eq <= 0) {
      throw new EnvParseError(
        "Not a valid env line (expected KEY=VALUE)",
        i + 1
      )
    }

    const key = trimmed.slice(0, eq).trim()
    if (!key) throw new EnvParseError("Empty variable key", i + 1)

    let value = trimmed.slice(eq + 1).trim()

    const quote = value[0]
    if (quote === '"' || quote === "'") {
      const end = value.lastIndexOf(quote)
      if (end <= 0) {
        throw new EnvParseError("Unclosed quote in value", i + 1)
      }
      value = value.slice(1, end)
    } else {
      const hash = value.indexOf(" #")
      if (hash !== -1) value = value.slice(0, hash).trim()
    }

    entries.push({ key, value })
  }

  return entries
}

/**
 * Build a Handlebars template from `.env`-style content, replacing each
 * variable's value with `{{KEY}}` so it resolves from the app's env vars at
 * render time. Comments and blank lines are preserved.
 */
export function envFileToTemplate(content: string): string {
  const out: string[] = []
  for (const raw of content.split(/\r?\n/)) {
    const trimmed = raw.trim()
    if (!trimmed || trimmed.startsWith("#")) {
      out.push(raw)
      continue
    }
    let head = trimmed
    let prefix = ""
    if (head.startsWith("export ")) {
      prefix = "export "
      head = head.slice("export ".length)
    }
    const eq = head.indexOf("=")
    if (eq <= 0) {
      out.push(raw)
      continue
    }
    const key = head.slice(0, eq).trim()
    out.push(`${prefix}${key}={{${key}}}`)
  }
  return out.join("\n") + "\n"
}

File: wails_backend/internal/lib/parseenv.go
package lib

import (
	"fmt"
	"strings"
)

type ParsedEnvEntry struct {
	Key   string
	Value string
}

type EnvParseError struct {
	Line    int
	Message string
}

func (e *EnvParseError) Error() string {
	return fmt.Sprintf("%s (line %d)", e.Message, e.Line)
}

func ParseEnvFile(content string) ([]ParsedEnvEntry, error) {
	var entries []ParsedEnvEntry
	lines := strings.Split(content, "\n")
	for i, line := range lines {
		line = strings.TrimRight(line, "\r")
		trimmed := strings.TrimSpace(line)
		if trimmed == "" || strings.HasPrefix(trimmed, "#") {
			continue
		}
		if strings.HasPrefix(trimmed, "export ") {
			line = trimmed[len("export "):]
			trimmed = strings.TrimSpace(line)
		}
		eq := strings.Index(trimmed, "=")
		if eq <= 0 {
			return nil, &EnvParseError{Message: "Not a valid env line (expected KEY=VALUE)", Line: i + 1}
		}
		key := strings.TrimSpace(trimmed[:eq])
		if key == "" {
			return nil, &EnvParseError{Message: "Empty variable key", Line: i + 1}
		}
		value := strings.TrimSpace(trimmed[eq+1:])
		if len(value) > 0 && (value[0] == '"' || value[0] == '\'') {
			quote := value[0]
			end := strings.LastIndex(value, string(quote))
			if end <= 0 {
				return nil, &EnvParseError{Message: "Unclosed quote in value", Line: i + 1}
			}
			value = value[1:end]
		} else if hash := strings.Index(value, " #"); hash != -1 {
			value = strings.TrimSpace(value[:hash])
		}
		entries = append(entries, ParsedEnvEntry{Key: key, Value: value})
	}
	return entries, nil
}

func EnvFileToTemplate(content string) string {
	var out []string
	for _, raw := range strings.Split(content, "\n") {
		raw = strings.TrimRight(raw, "\r")
		trimmed := strings.TrimSpace(raw)
		if trimmed == "" || strings.HasPrefix(trimmed, "#") {
			out = append(out, raw)
			continue
		}
		head := trimmed
		prefix := ""
		if strings.HasPrefix(head, "export ") {
			prefix = "export "
			head = head[len("export "):]
		}
		eq := strings.Index(head, "=")
		if eq <= 0 {
			out = append(out, raw)
			continue
		}
		key := strings.TrimSpace(head[:eq])
		out = append(out, prefix+key+"={{"+key+"}}")
	}
	return strings.Join(out, "\n") + "\n"
}
